crates/ra_hir_expand/src/diagnostics.rs

```diff
@@ -22,7 +22,12 @@ use crate::{db::AstDatabase, InFile};
 
 pub trait Diagnostic: Any + Send + Sync + fmt::Debug + 'static {
     fn message(&self) -> String;
+    /// A source to be used in highlighting and other visual representations
     fn source(&self) -> InFile<SyntaxNodePtr>;
+    /// A source to be used during the fix application
+    fn fix_source(&self) -> InFile<SyntaxNodePtr> {
+        self.source()
+    }
```

Comment:
Let's flip polarity here -- `source` should point to the node with problem, `highlight` should point to sub-range inside the `source`.

Reply:
And I *think* it should return an `Option`, as that's an API for implementor, not the API for user. User's API is in `impl dyn Diagnostic`

Reply:
> `highlight` should point to sub-range inside the `source`

The way it works currently for the `MissingFields*` diagnostics is that `source` and `highlight` don't intersect: the `source` is the `ast::RecordFieldList` (the list of the fields misses some elements) and the `highlight` is the `ast::Path`, both contained in the `ast::RecordLit`.

Reply:
I guess we should than use the whole literal (and not only field list) as `source`?

Reply:
Then I would love a brief into on what is `AstDiagnostic` is and on its relation to the `Diagnostic` in the context of the current changes.

Current `master` uses `Diagnostic::source` to derive the `AstDiagnostic::ast`
https://github.com/rust-analyzer/rust-analyzer/blob/e35804355797a50195718918efc6f7272786d62e/crates/ra_hir_ty/src/diagnostics.rs#L82-L90

that is later used to calculate a fix for the diagnostics.

I can sure change the logic to have three different pointers in the `ast::RecordLit`, but should I do that?

Reply:
It feels like `AstDiagnostic::ast` and `syntax_node` from `impl dyn Diagnostic` are very related and use `Diagnostic::source` in a similar fashion to derive the properly typed ast element to edit.

Can we merge all 3 methods into one?

Reply:
Deleting  `AstDiagnostic` for `BreakOutsideOfLoop`, `MismatchedArgCount`, `MissingUnsafe` caused no compile errors in `master`, so I'll go ahead and do some refactoring there, it might derive from your proposal a bit.

Reply:
https://github.com/SomeoneToIgnore/rust-analyzer/compare/less-red...SomeoneToIgnore:refactor-diagnostic-traits?expand=1#diff-6f6e28f1cd8bfc6335619935a7ccb9d1

Here's a draft with the following API:
```rust
pub trait Diagnostic: Any + Send + Sync + fmt::Debug + 'static {
    fn message(&self) -> String;
    fn source(&self) -> InFile<SyntaxNodePtr>;
    fn as_any(&self) -> &(dyn Any + Send + 'static);
    fn is_experimental(&self) -> bool {
        false
    }
}

pub trait DiagnosticsWithFixSource {
    type AST;
    fn fix_source(&self, db: &dyn AstDatabase) -> Option<Self::AST>;
}

impl dyn Diagnostic {
    pub fn downcast_ref<D: Diagnostic>(&self) -> Option<&D> {
        self.as_any().downcast_ref()
    }
}
```

but the range for the fix received from the `Self::AST.syntax()` is treated incorrectly.
`sema.original_range(fix_source.syntax()).range` panics on unknown node, despite the fact that we parse the files with into the same `db` in the code.

Seems like I cannot use the similar approach with the method in `sema` too due to the associated type limitations:

```rust
fn diagnostics_fix_range(&self, diagnostics_with_ast_source: &dyn DiagnosticsWithFixSource<AST=???>) -> FileRange
```

So far everything looks like I have to deal with this FIXME

https://github.com/rust-analyzer/rust-analyzer/blob/9da454bcc2386e2782b71a9756b77a7763fd544b/crates/ra_hir/src/semantics.rs#L665-L666

presumably in a separate PR, unless there's an easier way to deal with everything.

Reply:
Hm... I am wondering if perhaps a better move would be to customize ranges at the presentation layer? So that the core of a compile just flags "something went wrong around here", and then it's the IDE task to highlight the most reasonable bit, suggest a fix, etc....

Reply:
I've made a few various attempts to refactor the traits and here're the results:

1. Currently, the only custom fixes for diagnostics we have are located in the various `::on(` clauses here:
https://github.com/rust-analyzer/rust-analyzer/blob/e0de2475208765a171f335dfffde764f96243d41/crates/ra_ide/src/diagnostics.rs#L32

I think those diagnostics should be the only ones to implement the `AstDiagnostic` trait which purpose should be to provide a way to derive a fix out of the diagnostic data.

Since we care about the fixes on the ide level, I've decided to move the trait there and use `Fix` directly, with no AST associated param, `Optional` as a return value also included.
I've done a set of renames to separate various ranges one from each other better.

2. When determining the final assist ranges in `ra_ide`, we need to use `sema.diagnostics_range` and `sema.ast` that call `original_range` and also try to reparse the file for every diagnostic:
https://github.com/rust-analyzer/rust-analyzer/blob/e0de2475208765a171f335dfffde764f96243d41/crates/ra_hir/src/semantics.rs#L381

What was unexpected for me during the refactoring, is that even if I do the same reparse and derive an AST node from it, node's children will need an additional `sema.original_range` call before the range can be valid.
I had to use it once in the trait implementation:
https://github.com/rust-analyzer/rust-analyzer/blob/db12ccee96bf37367b39ad99638d06da7123c088/crates/ra_ide/src/diagnostics/diagnostics_with_fix.rs#L81

and I think there are more range errors dwell in the existing impls, so we need to investigate this further later.

This is the great test that helped me to catch this:
https://github.com/rust-analyzer/rust-analyzer/blob/e0de2475208765a171f335dfffde764f96243d41/crates/ra_ide/src/diagnostics.rs#L654

3. Previous two bullets result in a fact that we need to use `&SyntaxNode` in the base `ra_hir_expand::diagnostics::Diagnostic` trait instead of a more specific type: for diagnostics with no fixes, we have to determine the correct position for the highlighting and `sema::original_range` requires `&SyntaxNode`'s.
Additionally, the `sema::diagnostics_display_range` exists for the same purpose, but it's very error-prone, as described in the previous bullets, so we might need to deal with this sooner than later:

https://github.com/rust-analyzer/rust-analyzer/blob/9da454bcc2386e2782b71a9756b77a7763fd544b/crates/ra_hir/src/semantics.rs#L665-L666

4. Current way we use the `DiagnosticSink` had been left the same, but it had shrinked to this now:

https://github.com/rust-analyzer/rust-analyzer/blob/db12ccee96bf37367b39ad99638d06da7123c088/crates/ra_ide/src/diagnostics.rs#L54

It looks like the whole sink can be replaced with the `Vec` of `Diagnostic` implementations, so that we can try to cast each of them to the `DiagnosticWithFix` trait, but I failed to express it with Rust.